viewer-prototype/src/browser/trace-viewer/components/utils/filter-tree-component.tsx

```diff
@@ -0,0 +1,230 @@
+import * as React from 'react';
+import { Entry } from 'tsp-typescript-client/lib/models/entry';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faChevronDown, faChevronRight } from '@fortawesome/free-solid-svg-icons'
+import { CheckboxComponent } from './checkbox-component';
+
+type FilterTreeProps = {
+    nodes: Entry[];
+    checkedSeries: number[];
+    onChecked: (ids: number[]) => void;
+}
+
+type FilterTreeState = {
+    treeNodes: TreeNode[];
+    checkedIDs: number[];
+}
+
+export class FilterTree extends React.Component<FilterTreeProps, FilterTreeState> {
+    constructor(props: FilterTreeProps) {
+        super(props);
+        this.state = {
+            treeNodes: listToTree(this.props.nodes),
+            checkedIDs: this.props.checkedSeries
+        }
+    }
+
+    shouldComponentUpdate(nextProps: FilterTreeProps, nextState: FilterTreeState) {
+        return (this.props.checkedSeries !== nextProps.checkedSeries || this.props.nodes !== nextProps.nodes)
+        
+    }
+
+    componentWillReceiveProps(nextProps: FilterTreeProps) {
+        this.setState({
+            treeNodes: listToTree(nextProps.nodes),
+            checkedIDs: nextProps.checkedSeries
+        })
+    }
+
+    getRootNodes = () => {
+        const nodes = this.state.treeNodes;
+        return nodes.filter((node: TreeNode) => node.isRoot === true);
+    }
+
+    getChildNodes = (node: TreeNode) => {
+        if (!node.children) return [];
+        return node.children;
+    }
+
+    getNode: any = (node: TreeNode, id: number) => {
+        if (node.id === id) {
+            return node;
+        } else if (node.children !== undefined && node.children.length > 0) {
+            let result = null;
+            for (let i = 0; result == null && i < node.children.length; i++) {
+                result = this.getNode(node.children[i], id);
+            }
+            return result;
+        }
+        return null;
+    }
+
+    onToggle = (node: TreeNode) => {
+        const nodes = this.state.treeNodes;
+        const toggledNode = this.getNode(nodes[0], node.id);
```

Comment:
This should probably be a for loop, when we have experiments, there can be multiple root nodes

Also, you should maybe call a setState on some components if you really want it to work?